lib/route.js

```diff
@@ -297,10 +300,16 @@ function buildRouting (options) {
       if (ip !== undefined && req.headers['x-forwarded-host']) {
         hostname = req.headers['x-forwarded-host']
       }
+      proto = function () {
+        return req.headers['x-forwarded-proto']
+          .startsWith('https') ? 'https:' : 'http:'
+      }
     }
-
     var childLogger = logger.child({ [requestIdLogLabel]: req.id, level: context.logLevel })
     childLogger[kDisableRequestLogging] = disableRequestLogging
+    Object.defineProperty(req, 'protocol', {
```

Comment:
I think that is unnecessary because do not have any complex operation.
I give this solution if use regex.

Reply:
The goal to define this in the prototype is to avoid redefining the method here. Another option is to use the decorateRequest utility during startup, based on the trustProxy value.

Reply:
@mcollina 
I don’t get you, 
You suggest using decoratedRequest during the build ? 

```js
  fastify.decorateRequest('protocol', {
    getter: function () {
      if (options.trustProxy)
    ....
    }
  })
```

Reply:
after https://github.com/fastify/fastify/blob/1568dff519fcd0164c22187bf7e24456d83c75a5/fastify.js#L225, add:

```js
 if (options.trustProxy) {
   fastify.decorateRequest('protocol', {
     get: function () {
      ....
      }
  })
} else {
   fastify.decorateRequest('protocol', {
     get: function () {
      ....
      }
  })
}
```

Reply:
`getter` is correct:

```js
 if (options.trustProxy) {
   fastify.decorateRequest('protocol', {
     getter: function () {
      ....
      }
  })
} else {
   fastify.decorateRequest('protocol', {
     getter: function () {
      ....
      }
  })
}
```